designs/data_serialization_design.md

```diff
@@ -0,0 +1,189 @@
+  LevelSquashingFunction level_squash_fn = 9;
+  HostSquashingFunction host_squash_fn = 10;
+}
+
+message Bucket {
+  double probability = 1;
+  uint64 last_updated_time_ms = 2;
+}
+
+message Level {
+  repeated Bucket buckets = 1;
+}
+
+message FairData {
+  repeated Level levels = 1;
+}
+
+message FairRunConsts {
+  uint32 murmur_seed = 1;
+  string ref_guid = 2;
+}
+
+message HostMeta {
+  string host_guid = 1;
+  uint64 serialized_at_ms = 2;
+}
+
+message FairRunTimeData {
+  FairRunConsts runtime = 1;
+  FairData data = 2;
+}
+
+message FairSt {
+  // int32 version_num = 1; May be do this only if necessary
+  repeated TrackerCfg cfgs = 2;
+  repeated FairRunTimeData data = 3;
+  HostMeta meta = 4;
+}
+```
+
+The top-level struct contains a list of `TrackerCfgs`, the Runtime data is encoded using `FairRunTimeData` - which contains the consts and the data. Each unit of the run-time data contains a link `ref_guid` which refers to the config_guid and version. The structure also include a `HostMeta` field that includes information about the host.
+
+
+## Implementation Plan
+### Phase 1: Core
+
+Define schema (proto/fair/data/v1/structure.proto).
+Generate Go code with protoc-gen-go.
+Implement:
+```
+func (s *Structure) Serialize() ([]byte, error)
+func DeserializeStructure(data []byte) (*Structure, error)
+```
+
+### Phase 2: Integration
+- Add serialization tests
+- Validate deserialized data
+- Robust error handling
+- Benchmark performance 
+- Support human readable dumping (JSON)
+
+### Phase 3: Enhancements
+- Optional compression (gzip/lz4).
+- Checksums for integrity.
+- Migration utilities for schema upgrades.
+
+``` 
+type SerializationOptions struct {
+    Compress        bool
+    IncludeChecksum bool
+    SchemaVersion   uint32
+}
+
+func (s *Structure) Serialize(opts *SerializationOptions) ([]byte, error)
+func DeserializeStructure(data []byte) (*Structure, error)
+func DeserializeStructureWithOptions(data []byte, opts *DeserializationOptions) (*Structure, error)
+func ValidateStructure(s *Structure) error
+```
+
+Error handling includes structured errors (SerializationError) and standard cases (invalid data, unsupported schema, checksum mismatch, etc.).
+
+## Performance Considerations
+
+For a typical config (L=5, M=1000):
+- Raw: ~40KB (5 levels × 1000 buckets × 8 bytes).
+- With 100 trackers: ~4MB.
+
+Compression and deltas can reduce size further — to be validated with experiments.
+
+## Testing
+Correctness: Round-trip serialization tests. \
+Performance: Benchmarks for speed and memory use.
+
+## Monitoring & Observability
+Metrics: serialization latency, data size, error rates, schema usage.\
+Logging: events, errors, warnings (e.g., slow serialization).
+
+## Future Enhancements
+1) Compression.
+2) Delta syncs (share only updated fields).
+3) External cache (Redis) for global state.
+4) gRPC + service discovery for host coordination.
+5) Warm-up mechanisms for new hosts.
+
+## Thoughts on State Syncing
```

Comment:
Maybe we keep this one out for now and make that a separate design? This is super important but kind of a separate problem so want to separate.